### PR TITLE
Internal: Use workers to parallelize tests

### PR DESCRIPTION
**Pull request by yotamselementor**
See Playwright documentation:
* https://playwright.dev/docs/auth
* https://playwright.dev/docs/test-parallel#parallelize-tests-in-a-single-file
* https://playwright.dev/docs/test-fixtures

## PR Checklist

- [ ] The commit message follows our guidelines:  https://github.com/elementor/elementor/blob/master/.github/CONTRIBUTING.md

## PR Type
What kind of change does this PR introduce?

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] Other... Please describe:

## Summary

This PR can be summarized in the following changelog entry:

*

## Description
An explanation of what is done in this PR

*

## Test instructions
This PR can be tested by following these steps:

*

## Quality assurance

- [ ] I have tested this code to the best of my abilities
- [ ] I have added unittests to verify the code works as intended
- [ ] Docs have been added / updated (for bug fixes / features)

Fixes #

### CODE REVIEW

File: tests/elements-regression/playwright.config.ts [hunk 1]
```diff
@@ -1,32 +1,42 @@
-import { config as _config } from 'dotenv';
 import { resolve } from 'path';
 import { defineConfig } from '@playwright/test';
+import { config as _config } from 'dotenv';
+
+process.env.DEV_SERVER = 'http://localhost:8888';
+process.env.TEST_SERVER = 'http://localhost:8888'; // Test server is the same as dev server in elements-regression
+process.env.DEBUG_PORT = ( 1 === Number( process.env.TEST_PARALLEL_INDEX ) ) ? '9223' : '9222';
```
Comment (davseve) [inline tests/elements-regression/playwright.config.ts:7]: I saw this grouping parentheses in various places, they may be redundant.
```suggestion
process.env.DEBUG_PORT = 1 === Number( process.env.TEST_PARALLEL_INDEX ) ? '9223' : '9222';
```

File: tests/playwright/assets/api-requests.ts [hunk 2]
```diff
@@ -1,177 +1,139 @@
...
+			requests.push( request.delete( `${ this.baseUrl }/index.php`, {
+				headers: {
+					'X-WP-Nonce': this.nonce,
+				},
+				params: {
+					rest_route: `/wp/v2/media/${ ids[ id ] }`,
+					force: 1,
+				},
+			} ) );
+		}
```
Comment (davseve) [inline tests/playwright/assets/api-requests.ts:78]: Please wrap this block with empty lines.

File: tests/playwright/assets/wp-env-cli.ts [hunk 3]
```diff
@@ -1,18 +1,20 @@
...
+	if ( process.env.CI ) {
+		if ( logs.includes( 'Warning' ) ) {
+			github.warning( logs );
 		}
-	};
-}
+		if ( logs.includes( 'Error' ) ) {
+			github.error( logs );
+		}
+	}
```
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/assets/wp-env-cli.ts:17]: its not the same logic, we need to `return logs;` in all cases

File: tests/playwright/sanity/device-mode.test.ts [hunk 4]
```diff
@@ -1,9 +1,19 @@
...
 import WpAdminPage from '../pages/wp-admin-page';
 
 test.describe( 'Device mode', () => {
-	test( 'Correct device mode is returned on Desktop', async ( { page }, testInfo ) => {
-		const wpAdmin = new WpAdminPage( page, testInfo ),
+	test.beforeAll( async ( { browser, apiRequests }, testInfo ) => {
+		const page = await browser.newPage();
+		const wpAdmin = new WpAdminPage( page, testInfo, apiRequests );
+
+		await wpAdmin.resetExperiments();
```
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/sanity/device-mode.test.ts:10]: yep - just a different order of running

File: tests/elements-regression/playwright.config.ts [hunk 5]
```diff
@@ -1,32 +1,42 @@
-import { config as _config } from 'dotenv';
 import { resolve } from 'path';
 import { defineConfig } from '@playwright/test';
+import { config as _config } from 'dotenv';
+
+process.env.DEV_SERVER = 'http://localhost:8888';
+process.env.TEST_SERVER = 'http://localhost:8888'; // Test server is the same as dev server in elements-regression
```
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/elements-regression/playwright.config.ts]: Fixed: https://github.com/elementor/elementor/pull/27669/commits/27669005649038322722a4aa7f9c845c226b4a61

File: tests/playwright/assets/api-requests.ts [hunk 6]
```diff
@@ -1,177 +1,139 @@
...
-	if ( ! response.ok() ) {
-		throw new Error( `
-			Failed to create a ${ entity }: ${ response.status() }.
+		if ( ! response.ok() ) {
+			throw new Error( `
+			Failed to get a ${ entity }: ${ response.status() }.
 			${ await response.text() }
 		` );
+		}
+		return await response.json();
```
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/assets/api-requests.ts:116]: https://github.com/elementor/elementor/pull/27669/commits/f3bef0e4e41e8371a4f9f3dcf48dde2be0ab2904

File: tests/playwright/assets/wp-env-cli.ts [hunk 7]
```diff
@@ -1,18 +1,20 @@
...
-			}
-			if ( logs.includes( 'Error' ) ) {
-				github.error( logs );
-			}
+export const wpEnvCli = ( command: string ) => {
+	const cliCommand = ( 1 === Number( process.env.TEST_PARALLEL_INDEX ) ) ? 'tests-cli' : 'cli';
+	const cli = `npx wp-env run ${ cliCommand } `;
+	const logs = execSync( `${ cli }${ command }`, { encoding: 'utf-8' } );
+	// eslint-disable-next-line no-console
+	console.log( logs );
```
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/assets/wp-env-cli.ts:10]: Fixed: https://github.com/elementor/elementor/pull/27669/commits/4410f751df890cdb40024f0b40c01598d4fb463a

File: tests/playwright/assets/api-requests.ts (hunk 2 above)
Comment (yotamselementor) [reply tests/playwright/assets/api-requests.ts:79]: Fixed: https://github.com/elementor/elementor/pull/27669/commits/c46c0be3ca9d25f0b0eb82f5e26563fad49206ea

File: tests/elements-regression/playwright.config.ts (hunk 1 above)
Comment (yotamselementor) [reply tests/elements-regression/playwright.config.ts]: Fixed: https://github.com/elementor/elementor/pull/27669/commits/c46c0be3ca9d25f0b0eb82f5e26563fad49206ea

File: tests/elements-regression/playwright.config.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/elements-regression/playwright.config.ts:6]: Reverted: https://github.com/elementor/elementor/pull/27669/commits/3419c920418c0fd745c0d0d49d9ac40b1ef1b606

Apparently this is used here: https://github.com/elementor/elementor/blob/3419c920418c0fd745c0d0d49d9ac40b1ef1b606/tests/playwright/parallelTest.ts#L11

File: tests/playwright/assets/wp-env-cli.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (davseve) [reply tests/playwright/assets/wp-env-cli.ts:18]: We are returning logs
We still have the `return logs;` at the end of the function.

[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/assets/wp-env-cli.ts:18]: OK, so you mean this?

```typescript
	if ( ! process.env.CI ) {
		return logs;
	}

	if ( logs.includes( 'Warning' ) ) {
		github.warning( logs );
	}

	if ( logs.includes( 'Error' ) ) {
		github.error( logs );
	}

	return logs;
```

[the remark this answers is not in the document]
Comment (davseve) [reply tests/playwright/assets/wp-env-cli.ts:18]: Yes, that was my intention.

[the remark this answers is not in the document]
Comment (yotamselementor) [reply tests/playwright/assets/wp-env-cli.ts]: Fixed, thanks:
https://github.com/elementor/elementor/pull/27669/commits/2f026081d2a1fcaa0a11b9f2e16588b0a317b7bb

**Review by davseve — APPROVED** [review]

**Pull request merged**